### PR TITLE
T442: restore 'balance with confirmations' and 'balance at height' functionality

### PR DESCRIPTION
**Pull request by ismagin**
http://127.0.0.1:6869/debug/stateWaves/20

```{
  "address:3P9o3ZYwtHkaU1KxsKkFjJqJKS3dLHLC9oF" : 100000000,
  "address:3PAGPDPqnGkyhcihyjMHe9v36Y4hkAh9yDy" : 100000000,
  "address:3P8JdJGYc7vaLu4UXUZc1iRLdzrkGtdCyJM" : 100000000,
  "address:3PBWXDFUc86N2EQxKJmW8eFco65xTyMZx6J" : 100000000,
  "address:3PAWwWa6GbwcJaFzwqXQN5KQm7H96Y7SHTQ" : 9999999500000000,
  "address:3PJaDyprvekvPXPuAtxrapacuDJopgJRaU3" : 100000000
}```

### CODE REVIEW

File: src/main/scala/com/wavesplatform/state2/reader/StateReaderImpl.scala [hunk 1]
```diff
@@ -64,8 +54,14 @@ class StateReaderImpl(p: StateStorage) extends StateReader {
 
   override def accountPortfolios: Map[Account, Portfolio] =
     p.portfolios.entrySet().asScala
```
Comment (phearnot) [inline src/main/scala/com/wavesplatform/state2/reader/StateReaderImpl.scala]: No need to call entrySet:
```scala
    p.portfolios.asScala.map {
      case (acc, (b, (i, o), as)) =>
        Account.fromBytes(acc).explicitGet() ->
          Portfolio(b, LeaseInfo(i, o), as.map { case (k, v) => EqByteArray(k) -> v })
    }.toMap
```

File: src/main/scala/com/wavesplatform/state2/reader/StateReaderImpl.scala [hunk 2]
```diff
@@ -76,4 +72,10 @@ class StateReaderImpl(p: StateStorage) extends StateReader {
     .map(_.getKey)
     .map(EqByteArray)
     .toSeq
```
Comment (phearnot) [inline src/main/scala/com/wavesplatform/state2/reader/StateReaderImpl.scala]: While we're at it, please remove `entrySet` and replace filter/map with collect.

File: src/main/scala/com/wavesplatform/state2/reader/StateReader.scala [hunk 3]
```diff
@@ -108,5 +113,58 @@ object StateReader {
...
+    def stateHash(): Int = (BigInt(FastCryptographicHash(s.accountPortfolios.toString().getBytes)) % Int.MaxValue).toInt
+
+    private def minBySnapshot(acc: Account, atHeight: Int, confirmations: Int)(extractor: Snapshot => Long): Long = {
+      val bottomNotIncluded = atHeight - confirmations
+
+      @tailrec
+      def loop(deeperHeight: Int, list: Seq[Snapshot]): Seq[Snapshot] = {
+        if (deeperHeight == 0) list
+        else {
+          lazy val snapshot = s.snapshotAtHeight(acc, deeperHeight).get
```
Comment (phearnot) [inline src/main/scala/com/wavesplatform/state2/reader/StateReader.scala]: Does it make any sense to make `snapshot` a `lazy` val?

File: src/main/scala/com/wavesplatform/state2/diffs/BlockDiffer.scala [hunk 4]
```diff
@@ -38,22 +42,20 @@ object BlockDiffer extends ScorexLogging {
...
-          .filter { case (acc, portfolio) => portfolio.effectiveBalance != 0 }
-          .map { case (acc, portfolio) => (acc, s.accountPortfolio(acc).effectiveBalance, portfolio.effectiveBalance) }
-          .map { case (acc, oldEffBalance, effBalanceDiff) =>
-            EffectiveBalanceSnapshot(acc = acc,
-              height = s.height + 1,
-              prevEffectiveBalance = if (s.height == 0) (oldEffBalance + effBalanceDiff) else oldEffBalance,
-              effectiveBalance = oldEffBalance + effBalanceDiff)
+        val newSnapshots = diff.portfolios
+          .filter { case (acc, portfolioDiff) => portfolioDiff.balance != 0 || portfolioDiff.effectiveBalance != 0 }
+          .map { case (acc, portfolioDiff) =>
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/diffs/BlockDiffer.scala]: ok

File: src/main/scala/com/wavesplatform/state2/reader/StateReaderImpl.scala (hunk 1 above)
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/reader/StateReaderImpl.scala]: ok

File: src/main/scala/com/wavesplatform/state2/reader/StateReaderImpl.scala (hunk 2 above)
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/reader/StateReaderImpl.scala]: ok

File: src/main/scala/com/wavesplatform/state2/reader/StateReader.scala (hunk 3 above)
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/reader/StateReader.scala]: ok, not anymore

**Pull request merged**